File: Strivers-A2Z-DSA-Sheet/06.Bit Manipulation/1. Learn Bit Manipulation/python/05.Set the righmost unset bit.py
"""QUESTION:
Given a non-negative integer `N`, set its rightmost unset bit. If no unset bit exists, return `N` unchanged.

Approach:
If `N + 1` is a power of two, then every lower bit in `N` is already set, so the number stays the same. Otherwise, `N | (N + 1)` sets the rightmost zero bit.

Time Complexity: O(1)
Space Complexity: O(1)
"""

def is_power_of_two(n: int) -> bool:
    if n <= 0:
        return False
    return (n & (n - 1)) == 0

def set_bit(N: int) -> int:
    if is_power_of_two(N + 1):
        return N
    return N | (N + 1)

File: Strivers-A2Z-DSA-Sheet/06.Bit Manipulation/1. Learn Bit Manipulation/javascript/05.Set the righmost unset bit.js
/**
QUESTION:
Given a non-negative integer `N`, set its rightmost unset bit. If no unset bit exists, return `N` unchanged.

Approach:
If `N + 1` is a power of two, then `N` already has all lower bits set, so return `N`. Otherwise, `N | (N + 1)` sets the rightmost zero bit.

Time Complexity: O(1)
Space Complexity: O(1)
*/

function isPowerOfTwo(n) {
    if (n <= 0) return false;
    return (n & (n - 1)) === 0;
}

function setBit(N) {
    if (isPowerOfTwo(N + 1)) return N;
    return N | (N + 1);
}

module.exports = { setBit };
